dot/state/hashtoblockmap_test.go

```diff
@@ -0,0 +1,327 @@
+						0x26, 0x36, 0x30, 0x28, 0x11, 0x5a, 0x9f, 0x2c,
+						0xf9, 0x63, 0xee, 0xed, 0x52, 0x6b, 0x8b, 0x33,
+						0xe5, 0x31, 0x6f, 0x8, 0xb5, 0xd, 0xd, 0xc3,
+					}: {Header: types.Header{Number: big.NewInt(99)}},
+				},
+			},
+			block: &types.Block{Header: types.Header{Number: big.NewInt(1)}},
+			expectedMap: &hashToBlockMap{
+				mapping: map[common.Hash]*types.Block{
+					{
+						0x64, 0x43, 0xa0, 0xb4, 0x6e, 0x4, 0x12, 0xe6,
+						0x26, 0x36, 0x30, 0x28, 0x11, 0x5a, 0x9f, 0x2c,
+						0xf9, 0x63, 0xee, 0xed, 0x52, 0x6b, 0x8b, 0x33,
+						0xe5, 0x31, 0x6f, 0x8, 0xb5, 0xd, 0xd, 0xc3,
+					}: {Header: headerWithHash(types.Header{Number: big.NewInt(1)})},
+				},
+			},
+		},
+		"store new block": {
+			initialMap: &hashToBlockMap{
+				mapping: map[common.Hash]*types.Block{},
+			},
+			block: &types.Block{Header: types.Header{Number: big.NewInt(1)}},
+			expectedMap: &hashToBlockMap{
+				mapping: map[common.Hash]*types.Block{
+					{
+						0x64, 0x43, 0xa0, 0xb4, 0x6e, 0x4, 0x12, 0xe6,
+						0x26, 0x36, 0x30, 0x28, 0x11, 0x5a, 0x9f, 0x2c,
+						0xf9, 0x63, 0xee, 0xed, 0x52, 0x6b, 0x8b, 0x33,
+						0xe5, 0x31, 0x6f, 0x8, 0xb5, 0xd, 0xd, 0xc3,
+					}: {Header: headerWithHash(types.Header{Number: big.NewInt(1)})},
+				},
+			},
+		},
+	}
+
+	for name, testCase := range testCases {
+		testCase := testCase
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			htb := testCase.initialMap
+
+			htb.store(testCase.block)
+
+			assert.Equal(t, testCase.expectedMap, htb)
+		})
+	}
+}
+
+func Test_hashToBlockMap_delete(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		initialMap    *hashToBlockMap
+		hash          common.Hash
+		deletedHeader *types.Header
+		expectedMap   *hashToBlockMap
+	}{
+		"hash does not exist": {
+			initialMap: &hashToBlockMap{
+				mapping: map[common.Hash]*types.Block{},
+			},
+			hash: common.Hash{1, 2, 3},
+			expectedMap: &hashToBlockMap{
+				mapping: map[common.Hash]*types.Block{},
+			},
+		},
+		"hash deleted": {
+			initialMap: &hashToBlockMap{
+				mapping: map[common.Hash]*types.Block{
+					{1, 2, 3}: {Header: types.Header{ParentHash: common.Hash{1, 2, 3}}},
+				},
+			},
+			hash:          common.Hash{1, 2, 3},
+			deletedHeader: &types.Header{ParentHash: common.Hash{1, 2, 3}},
+			expectedMap: &hashToBlockMap{
+				mapping: map[common.Hash]*types.Block{},
+			},
+		},
+	}
+
+	for name, testCase := range testCases {
+		testCase := testCase
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			htb := testCase.initialMap
+
+			deletedHeader := htb.delete(testCase.hash)
+
+			assert.Equal(t, testCase.deletedHeader, deletedHeader)
+			assert.Equal(t, testCase.expectedMap, htb)
+		})
+	}
+}
+
+func Test_hashToBlockMap_threadSafety(t *testing.T) {
+	// This test consists in checking for concurrent access
+	// using the -race detector.
+	t.Parallel()
+
+	var startWg, endWg sync.WaitGroup
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
why do we need this context?

Reply:
To cancel all `runInLoop` goroutines all at once.

```go
		for ctx.Err() == nil {
			f()
		}
```

and the cancel is called by the timer.

To be fair, all this feels like over-engineering but also like the right way to test for data races *quickly*, since it has to be 'tight' and as event-driven as possible

Reply:
I don't really see the need to use a context to handle cancellation.  Could you not just close some channel to signify cancellation?

Reply:
Yes we can use a struct{} channel that gets closed, it's just that it then becomes:

```go
stop := false
for !stop {
	select {
	case <-stopGoroutines:
		stop = true
	default:
	}
	f()
}
```

Which is longer/more confusing I think.